### ISSUE TITLE
Feature request: Option to specify a specific plugin module name when prefixed with a ~

### ISSUE THREAD

**Issue by impinball** [eslint/eslint#2796]
Initially proposed in [babel/babel-eslint#131 (comment)](https://github.com/babel/babel-eslint/issues/131#issuecomment-113330418)

Custom parsers and custom named modules should be able to be included as plugins. So for example, with something like babel-eslint, the parser and additional rules could be housed in the same project instead of separate repos, therefore, keeping babel-eslint related issues consolidated to a single repo. An example configuration would be (assuming custom rules were prefixed with a `~`):

```json
{
  "parser": "babel-eslint",
  "plugins": [
    "~babel-eslint/plugin"
  ],
  "rules": {
    "babel-eslint/es7.objectRestSpread": 1
  }
}
```

(Copied + some editing from PR #2793)

**Comment by nzakas** [comment]
It seems like the title and description of this issue aren't quite the same. Also, you're conflating a problem statement with a solution, so it's a little bit hard to follow. Let's start from the top:

What is the problem you are trying to solve?

**Comment by mtscout6** [comment]
The problem is to allow a custom parser also provide rules for the variant of the language being parsed.

**Comment by nzakas** [comment]
Once again, you are describing the solution to a problem, but not the problem itself.

If I may infer the problem, it sounds like you want the ability to have a single npm package containing both a parser and rules. Your suggested solution to that problem is that parsers should be able to provide rules. I dont think such a solution makes sense because that's not the role of a parser.

However, I do believe it is already possible to bundle a parser with a plugin. For instance:

```
plugins: [ "foo" ]
parser: "eslint-plugin-foo/parser"
```

I'm pretty sure this already works .

**Comment by MoOx** [comment]
What we would like to see is

```
parser: "babel-eslint"
plugins: [ "babel-eslint/plugin" ]
```

If we have the corresponding file `node_modules/babel-eslint/plugins.js`, is this going to work (I guess nobody of us tried yet). Because I think that we think that only npm module names work in the `plugins` array.

**Comment by mtscout6** [comment]
@nzakas Thank you for helping me to understand your library and what it can do. I was unaware that such was possible.

I apologize for not wording the problem in a format that was appealing to you, at the time I didn't see that my use of the word "allow" was attempting to solve the problem. That was my attempt to word the problem as you requested.

With that I will walk away from the situation. I don't appreciate trying to search out a solution for a problem that provides a better downstream user experience, only to receive backlash and public shaming for doing so. We are all problem solvers working to provide a better experience for our customers, fellow colleagues, and also ourselves. I would hope that others like myself can see it in that order, though it cannot be imposed.

**Comment by nzakas** [comment]
@mtscout6 I always ask for people to describe problems without mentioning the solution, as that reveals a lot more about the request. In fact, this separation of problem and solution is explicitly mentioned in our contribution guidelines for enhancements: http://eslint.org/docs/developer-guide/contributing#filing-issues

Please keep in mind, I spend a significant portion of the day triaging issues. Many people come to try to solve their own problem and my (and the team's) job is to try to distill those into generalizable problems and solutions. I've tried to foster this community in such a way that we look for reasons to say "yes" to requests, but ultimately, there are many times we end up saying "no" for one reason or another. What you will see consistently in our issues, however, is an honest attempt to really understand the ask before saying yes or no.

In this case, the discussion started on a different repo on an issue that is somewhat unrelated to this discussion. Then, it progressed to a pull request without much context (once again, or contribution guidelines are pretty clear that PRs require an issue). My only goal was to get the conversation back on track in adherence with this project's development process.

I'll admit that I was quite frustrated with this process and could have been more patient throughout. For that, I sincerely apologize. It was never my intention to publicly shame anyone, and I will use more care with the words I choose in the future.

**Comment by nzakas** [comment]
@MoOx that is not possible today, and pretty much not possible in the long run. Plugins double as namespaces, such as `react/some-rule`. That namespace is determined from the plugin package name.

I'm really against having parser packages do more than provide a parser. Making parsers into pseudoplugins means having two different ways of doing the same thing, and that almost always is a bad design decision. The decision to keep parsers, plugins, and configs separate was intentional and based on the existing JS parser ecosystem. 

However, I do think plugins already solve the core problem of having a single package bundle rules and parser. Do you want to try what I mentioned and see if that addresses your need?

**Comment by MoOx** [comment]
We will see :)

According to what you ask, I think we can close this issue for now.
